alchemy/src/aws/bucket.ts

```diff
@@ -127,108 +114,89 @@ export interface Bucket extends Resource<"s3::Bucket">, BucketProps {
+            `<Tag><Key>${Key}</Key><Value>${Value}</Value></Tag>`,
+        )
+        .join("")}</TagSet></Tagging>`;
+    };
+
+    // Try to check if bucket exists and update tags if needed
+    const bucketExists = yield* client
+      .request("HEAD", `/${props.bucketName}`)
+      .pipe(
+        Effect.map(() => true),
+        Effect.catchSome((err) =>
+          err instanceof AwsResourceNotFoundError
+            ? Effect.succeed(false)
+            : Effect.fail(err),
         ),
       );
 
+    if (bucketExists) {
       // Update tags if they changed and bucket exists
       if (this.phase === "update" && props.tags) {
-        await retry(() =>
-          client.send(
-            new PutBucketTaggingCommand({
-              Bucket: props.bucketName,
-              Tagging: {
-                TagSet: Object.entries(props.tags!).map(([Key, Value]) => ({
-                  Key,
-                  Value,
-                })),
-              },
-            }),
-          ),
-        );
+        const taggingXml = createTaggingXml(props.tags);
+        yield* client.put(`/${props.bucketName}?tagging`, taggingXml, {
+          "Content-Type": "application/xml",
+        });
       }
-    } catch (error: any) {
-      if (error.name === "NotFound") {
-        // Create bucket if it doesn't exist
-        await retry(() =>
-          client.send(
-            new CreateBucketCommand({
-              Bucket: props.bucketName,
-              // Add tags during creation if specified
-              ...(props.tags && {
-                Tagging: {
-                  TagSet: Object.entries(props.tags).map(([Key, Value]) => ({
-                    Key,
-                    Value,
-                  })),
-                },
-              }),
-            }),
-          ),
-        );
-      } else {
-        throw error;
+    } else {
+      // Create bucket if it doesn't exist
+      yield* client.put(`/${props.bucketName}`);
+
+      // Add tags after creation if specified
+      if (props.tags) {
+        const taggingXml = createTaggingXml(props.tags);
+        yield* client.put(`/${props.bucketName}?tagging`, taggingXml, {
+          "Content-Type": "application/xml",
+        });
       }
     }
 
-    // Get bucket details
+    // Get bucket details in parallel
     const [locationResponse, versioningResponse, aclResponse] =
-      await Promise.all([
-        retry(() =>
-          client.send(
-            new GetBucketLocationCommand({ Bucket: props.bucketName }),
-          ),
-        ),
-        retry(() =>
-          client.send(
-            new GetBucketVersioningCommand({ Bucket: props.bucketName }),
-          ),
-        ),
-        retry(() =>
-          client.send(new GetBucketAclCommand({ Bucket: props.bucketName })),
-        ),
+      yield* Effect.all([
+        client.get(`/${props.bucketName}?location`),
+        client.get(`/${props.bucketName}?versioning`),
+        client.get(`/${props.bucketName}?acl`),
```

Comment:
The docs suggest we might need a header?

GET /?location HTTP/1.1
Host: Bucket.s3.amazonaws.com

https://docs.aws.amazon.com/AmazonS3/latest/API/API_GetBucketLocation.html